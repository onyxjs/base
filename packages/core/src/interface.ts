import Suite, { isSuite, rootSymbol } from './suite';
import Test from './test';

const root = new Suite('root', { skip: false, todo: false }, null);
root[rootSymbol] = true;
let currentRoot: Suite = root;

export const getCurrentRoot: () => Suite = () => currentRoot;
export const setCurrentRoot: (v: Suite) => void = (v: Suite) => {
  if (!isSuite(v)) { return; }
  currentRoot = v;
};

export { root, currentRoot };

// istanbul ignore next internal
function _it(description: string, fn: () => void, skip = false, todo = false): Test {
  if (currentRoot.isRoot()) {
    throw new Error(`"${description}" "it" should not be called outside of "describe" block`);
  }

  const test = new Test(
    description,
    fn,
    {
      skip,
      todo,
    },
    currentRoot,
  );

  currentRoot.addChildren(test);
  return test;
}

export type ItFn = (description: string, fn: () => void) => Test;
// tslint:disable-next-line:interface-name
export interface It extends ItFn {
  skip: ItFn;
  todo: ItFn;
}

const it: It = ((description: string, fn: () => void) => _it(description, fn, false, false)) as It;
it.skip = (description: string, fn: () => void) => _it(description, fn, true, false);
it.todo = (description: string, fn: () => void) => _it(description, fn, false, true);

// istanbul ignore next internal
function _describe(description: string, fn: () => void, skip = false, todo = false): Suite {
  const suite = new Suite(
    description,
    {
      skip,
      todo,
    },
    currentRoot,
  );

  currentRoot = suite;
  fn();
  if (!suite.options.skip && !suite.options.todo && suite.children.length <= 0) {
    console.warn(`suite "${suite.getFullDescription()}" doesn't have any child tests or suites.`);
  }
  currentRoot = currentRoot.parent || root;

  currentRoot.addChildren(suite);
  return suite;
}

export type DescribeFn = (description: string, fn: () => void) => Suite;
// tslint:disable-next-line:interface-name
export interface Describe extends DescribeFn {
  skip: DescribeFn;
  todo: DescribeFn;
}

const describe: Describe = ((description: string, fn: () => void) =>
  _describe(description, fn, false, false)) as Describe;
describe.skip = (description: string, fn: () => void) => _describe(description, fn, true, false);
describe.todo = (description: string, fn: () => void) => _describe(description, fn, false, true);

export { it, describe };
export { beforeAll, afterAll, beforeEach, afterEach } from './hooks';
