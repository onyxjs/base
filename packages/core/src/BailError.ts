export class BailError extends Error {
  /* istanbul ignore next */
  constructor(message: string) {
    super(message)
    this.name = 'BailError'
  }
}
