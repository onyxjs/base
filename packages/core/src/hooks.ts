import { currentRoot } from './interface'

export type HookFn = (() => void) | (() => Promise<void>)
export type Hook = HookFn[]
export interface Hooks {
  beforeAll: Hook
  afterAll: Hook
  beforeEach: Hook
  afterEach: Hook
}
export type HookName = keyof Hooks

export const beforeAll: (...hooks: HookFn[]) => number = (...hooks: HookFn[]) => currentRoot.hooks.beforeAll.push(...hooks)
export const afterAll: (...hooks: HookFn[]) => number = (...hooks: HookFn[]) => currentRoot.hooks.afterAll.push(...hooks)
export const beforeEach: (...hooks: HookFn[]) => number = (...hooks: HookFn[]) => currentRoot.hooks.beforeEach.push(...hooks)
export const afterEach: (...hooks: HookFn[]) => number = (...hooks: HookFn[]) => currentRoot.hooks.afterEach.push(...hooks)
