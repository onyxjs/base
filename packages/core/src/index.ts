export {
  root,
  currentRoot,
  ItFn,
  It,
  it,
  DescribeFn,
  Describe,
  describe,
  getCurrentRoot,
  setCurrentRoot,
  beforeAll,
  afterAll,
  beforeEach,
  afterEach,
} from './interface'
export { Status, default as Result } from './result'
export { default as Runnable, isRunnable, RunnableTypes } from './runnable'
export { default as Test, isTest } from './test'
export { default as Suite, isSuite, SuiteStats, rootSymbol, BailError } from './suite'
export { default as Runner, RunOptions, runnerDefaults } from './runner'
export { Hook, HookName, HookFn, Hooks } from './hooks'
