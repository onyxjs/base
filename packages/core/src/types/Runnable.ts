export enum RunnableTypes {
  Runnable = 'runnable',
  Suite = 'suite',
  Test = 'test',
}

export interface RunnableOptions {
  skip: boolean
  todo: boolean
}
