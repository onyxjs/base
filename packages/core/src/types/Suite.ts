export interface SuiteStats {
  total: number
  pending: number
  running: number
  done: number
  passed: number
  failed: number
  skipped: number
  todo: number
  time: number
}
