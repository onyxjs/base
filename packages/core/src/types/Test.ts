export type TestFn = () => (void | Promise<any>)
