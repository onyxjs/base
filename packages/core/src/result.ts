// import { RunnableTypes } from "."\


// export enum Status {
//   Pending = 'pending',
//   Running = 'running',
//   Passed = 'passed',
//   Failed = 'failed',
//   Skipped = 'skipped',
//   Todo = 'todo',
// }

// Types
import { BailError } from './BailError'
import { TimeoutError } from './TimeoutError'
import { RunnableTypes, Status } from './types'

/**
 * @todo Delete messages.
 */
export default class Result {
  private _internalErrors: Error[]
  private _internalStatus: Status
  private _internalMessages: string[]
  
  public time: number
  public title: string
  public description: string
  public type: RunnableTypes

  constructor(messages: string | string[] = [], options: Result, errors: Error[] | Error = [], status?: Status) {
    this._internalErrors = !Array.isArray(errors) ? [ errors ] : errors
    this._internalStatus = status || Status.Pending
    this._internalMessages = !Array.isArray(messages) ? [ messages ] : messages
    
    this.description = options.description
    this.time = options.time
    this.title = options.title
    this.type = options.type
  }

  /**
   * @description Checks if the internal status is 'Pending' or 'Running'.
   */
  public isDone(): boolean {
    return this._internalStatus !== Status.Pending && this._internalStatus !== Status.Running
  }

  public get errors(): Array<Error> {
    return this._internalErrors
  }

  /**
   * @description Gets the internal status on the current `Result` instance.
   */
  public get status(): Status {
    return this._internalStatus
  }

  /**
   * @description Sets the internal status on the current `Result` instance.
   */
  public set status(v: Status) {
    if (this.isDone()) { return }
    this._internalStatus = v
  }

  /**
   * @description Gets the internal messages on the current `Result` instance.
   */
  public get messages(): string[] {
    return this._internalMessages
  }

  /**
   * @description Adds messages to the internal messages if the `Runnable` has not completed.
   */
  public addMessages(...messages: string[]): void {
    if (this.isDone()) { return }
    this._internalMessages = [...this._internalMessages, ...messages]
  }

  public addErrors(...errors: Error[]): void {
    if (this.isDone()) { return }
    this.addMessages(...errors.map((e) => e.message))
    this._internalErrors = [...this._internalErrors, ...errors]
  }
}
