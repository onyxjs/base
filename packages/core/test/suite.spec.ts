import { Status } from '../src/result';
import Runnable from '../src/runnable';
import Suite, { isSuite, rootSymbol } from '../src/suite';
import Test from '../src/test';

describe('Suite', () => {
  const defaultOpts = {
    skip: false,
    todo: false,
  };

  it('should run children', () => {
    const parent = new Suite('parent', defaultOpts, null);
    const child = new Suite('child', defaultOpts, parent);
    const spy = jest.spyOn(child, 'run');
    parent.addChildren(child);

    const start = jest.fn();
    parent.on('start', start);
    const pass = jest.fn();
    parent.on('pass', pass);
    const end = jest.fn();
    parent.on('end', end);
    parent.run();

    expect(spy).toHaveBeenCalled();

    expect(start).toHaveBeenCalledTimes(1);
    expect(pass).toHaveBeenCalledTimes(1);
    expect(end).toHaveBeenCalledTimes(1);
  });

  it('should skip', () => {
    const suite = new Suite('parent', { skip: true }, null);

    const start = jest.fn();
    suite.on('start', start);
    const skip = jest.fn();
    suite.on('skip', skip);
    const end = jest.fn();
    suite.on('end', end);
    suite.run();

    expect(start).toHaveBeenCalledTimes(0);
    expect(skip).toHaveBeenCalledTimes(1);
    expect(skip).toHaveBeenCalledWith(suite, false);
    expect(end).toHaveBeenCalledTimes(1);
  });

  it('should work with todo option', () => {
    const suite = new Suite('parent', { todo: true }, null);

    const start = jest.fn();
    suite.on('start', start);
    const skip = jest.fn();
    suite.on('skip', skip);
    const end = jest.fn();
    suite.on('end', end);
    suite.run();

    expect(start).toHaveBeenCalledTimes(0);
    expect(skip).toHaveBeenCalledTimes(1);
    expect(skip).toHaveBeenCalledWith(suite, true);
    expect(end).toHaveBeenCalledTimes(1);
  });

  it('should check if is root', () => {
    const suite = new Suite('suite', defaultOpts, null);

    expect(suite.isRoot()).toBeFalsy();

    suite[rootSymbol] = true;
    expect(suite.isRoot()).toBeTruthy();
  });

  it('should push children to children array', () => {
    const suite = new Suite('Suite', defaultOpts, null);
    const child1 = new Suite('desc', defaultOpts, suite);
    const child2 = new Suite('desc', defaultOpts, suite);

    expect(suite.children).toEqual([]);

    suite.addChildren(child1, child2);
    expect(suite.children.length).toEqual(2);
  });

  // tslint:disable-next-line:max-classes-per-file
  class PassingRunnable extends Runnable {
    public run() {
      this.result.addMessages('OK');
      this.result.status = Status.Passed;
      return this.result;
    }
  }
  it('should pass and collect messages', () => {
    const suite = new Suite('Suite', defaultOpts, null);
    const child = new PassingRunnable('desc', defaultOpts, null);
    suite.addChildren(child);

    const start = jest.fn();
    suite.on('start', start);
    const pass = jest.fn();
    suite.on('pass', pass);
    const end = jest.fn();
    suite.on('end', end);
    suite.run();

    expect(suite.result).toMatchSnapshot();

    expect(start).toHaveBeenCalledTimes(1);
    expect(pass).toHaveBeenCalledTimes(1);
    expect(end).toHaveBeenCalledTimes(1);
  });

  // tslint:disable-next-line:max-classes-per-file
  class FailingRunnable extends Runnable {
    public run() {
      this.result.addMessages('FAIL!');
      this.result.status = Status.Failed;
      return this.result;
    }
  }
  it('should fail and collect messages', () => {
    const suite = new Suite('Suite', defaultOpts, null);
    const child = new FailingRunnable('desc', defaultOpts, suite);
    suite.addChildren(child);

    const start = jest.fn();
    suite.on('start', start);
    const fail = jest.fn();
    suite.on('fail', fail);
    const end = jest.fn();
    suite.on('end', end);
    suite.run();

    expect(suite.result).toMatchSnapshot();

    expect(start).toHaveBeenCalledTimes(1);
    expect(fail).toHaveBeenCalledTimes(1);
    expect(end).toHaveBeenCalledTimes(1);
  });

  it('should not bail out', () => {
    const suite = new Suite('Suite', defaultOpts, null);
    const child1 = new FailingRunnable('desc', defaultOpts, suite);
    const child2 = new PassingRunnable('desc', defaultOpts, suite);
    suite.addChildren(child1, child2);

    suite.run();
    expect(suite.getStats().done).toBe(2);
  });

  it('should bail out when bail is set to true', () => {
    const suite = new Suite('Suite', defaultOpts, null);
    const child1 = new FailingRunnable('desc', defaultOpts, suite);
    const child2 = new PassingRunnable('desc', defaultOpts, suite);
    suite.addChildren(child1, child2);

    suite.run({ bail: true, timeout: 0, sequential: false });
    expect(suite.getStats().done).toBe(1);
  });

  it('should bail out after n tests when bail is a number', () => {
    const suite = new Suite('Suite', defaultOpts, null);
    const child = new PassingRunnable('desc', defaultOpts, suite);
    const child1 = new FailingRunnable('desc', defaultOpts, suite);
    const child2 = new FailingRunnable('desc', defaultOpts, suite);
    const child3 = new FailingRunnable('desc', defaultOpts, suite);
    const child4 = new PassingRunnable('desc', defaultOpts, suite);
    suite.addChildren(child, child1, child2, child3, child4);

    suite.run({ bail: 2, timeout: 0, sequential: false });
    expect(suite.getStats().done).toBe(3);
  });

  it('should check if is suite', () => {
    expect(isSuite(null)).toBeFalsy();
    expect(isSuite({})).toBeFalsy();
    expect(isSuite(new Runnable('not a suite', defaultOpts, null))).toBeFalsy();
    expect(isSuite(new Test('not a suite', () => null, defaultOpts, null))).toBeFalsy();
    expect(isSuite(new Suite('a suite', defaultOpts, null))).toBeTruthy();
  });

  it('should collect stats', () => {
    const passing = new Suite('passing', defaultOpts, null);
    passing.addChildren(
      new PassingRunnable('passing runnable', defaultOpts, null),
    );
    const skipped = new Suite('skipped', { skip: true }, null);
    const todo = new Suite('todo', { todo: true }, null);
    const failing = new Test('failing', () => {
      throw new Error('FAIL!');
    }, defaultOpts, null);
    const parent = new Suite('parent', defaultOpts, null);
    parent.addChildren(passing, skipped, todo, failing);

    expect(parent.getStats()).toMatchSnapshot({
      time: expect.any(Number),
    });

    parent.run();

    expect(parent.getStats()).toMatchSnapshot({
      time: expect.any(Number),
    });
  });

  it('should invoke hooks', () => {
    const error = console.error;
    console.error = jest.fn();

    const parent = new Suite('parent', {}, null);
    parent.addChildren(
      new Test('passing', () => null, {}, parent),
      new Test('failing', () => { throw new Error('Fail'); }, {}, parent),
    );
    const calls: string[] = [];

    parent.hooks.beforeAll.push(
      () => calls.push('beforeAll1'),
      () => calls.push('beforeAll2'),
    );
    parent.hooks.beforeEach.push(
      () => calls.push('beforeEach1'),
      () => calls.push('beforeEach2'),
      () => { throw new Error('beforeEach hook error'); },
    );
    parent.hooks.afterEach.push(
      () => calls.push('afterEach1'),
      () => calls.push('afterEach2'),
    );
    parent.hooks.afterAll.push(
      () => calls.push('afterAll1'),
      () => calls.push('afterAll2'),
      () => { throw new Error('afterAll hook error'); },
    );

    parent.run();
    expect(calls).toMatchSnapshot();

    expect(console.error).toHaveBeenCalledTimes(3);
    expect(console.error).toHaveBeenCalledWith('Error in beforeEach hook: Error: beforeEach hook error');
    expect(console.error).toHaveBeenCalledWith('Error in afterAll hook: Error: afterAll hook error');

    console.error = error;
  });

  describe('async', () => { // TODO: check this for races
    it('should pass', async () => {
      const child = new Test('child', () => null, defaultOpts, null);
      const parent = new Suite('parent', defaultOpts, null);
      parent.addChildren(child);

      const start = jest.fn();
      parent.on('start', start);
      const pass = jest.fn();
      parent.on('pass', pass);
      const end = jest.fn();
      parent.on('end', end);

      const promise = parent.asyncRun();

      expect(start).toHaveBeenCalledTimes(1);

      expect((await promise).status).toBe(Status.Passed);
      expect(pass).toHaveBeenCalledTimes(1);
      expect(end).toHaveBeenCalledTimes(1);
    });

    it('should run sequentially', () => {
      const suite = new Suite('Suite', defaultOpts, null);
      const child = new PassingRunnable('desc', defaultOpts, suite);
      const child1 = new PassingRunnable('desc', defaultOpts, suite);
      const child2 = new PassingRunnable('desc', defaultOpts, suite);
      const child3 = new PassingRunnable('desc', defaultOpts, suite);
      const child4 = new PassingRunnable('desc', defaultOpts, suite);
      suite.addChildren(child, child1, child2, child3, child4);

      suite.run({ bail: 2, timeout: 0, sequential: true });
      expect(suite.getStats().done).toBe(5);
    });

    it('should fail', async () => {
      const fn = jest.fn();

      const err = new Error('FAIL!');
      const child = new Test('child 1', () => { throw err; }, defaultOpts, null);
      const passingChild = new Test('child 2', fn, defaultOpts, null);
      const parent = new Suite('parent', defaultOpts, null);
      parent.addChildren(child);
      parent.addChildren(passingChild);

      const start = jest.fn();
      parent.on('start', start);
      const fail = jest.fn();
      parent.on('fail', fail);
      const end = jest.fn();
      parent.on('end', end);

      expect((await parent.asyncRun()).status).toBe(Status.Failed);

      expect(fn).toHaveBeenCalledTimes(1);
      expect(start).toHaveBeenCalledTimes(1);
      expect(fail).toHaveBeenCalledTimes(1);
      expect(end).toHaveBeenCalledTimes(1);
    });

    it.todo('should bail out');

    it('should skip', async () => {
      const parent = new Suite('parent', { skip: true }, null);

      const start = jest.fn();
      parent.on('start', start);
      const skip = jest.fn();
      parent.on('skip', skip);
      const end = jest.fn();
      parent.on('end', end);

      const promise = parent.asyncRun();

      expect(skip).toHaveBeenCalledTimes(1);
      expect(end).toHaveBeenCalledTimes(1);
      expect((await promise).status).toBe(Status.Skipped);
    });

    it('should invoke hooks', async () => {
      const error = console.error;
      console.error = jest.fn();

      const parent = new Suite('parent', {}, null);
      parent.addChildren(
        new Test('passing', () => null, {}, parent),
        new Test('failing', () => { throw new Error('Fail'); }, {}, parent),
      );
      const calls: string[] = [];

      parent.hooks.beforeAll.push(
        () => calls.push('beforeAll1'),
        () => calls.push('beforeAll2'),
      );
      parent.hooks.beforeEach.push(
        async () => await calls.push('beforeEach1'),
        () => calls.push('beforeEach2'),
        () => { throw new Error('beforeEach hook error'); },
      );
      parent.hooks.afterEach.push(
        () => calls.push('afterEach1'),
        () => calls.push('afterEach2'),
      );
      parent.hooks.afterAll.push(
        () => calls.push('afterAll1'),
        async () => await calls.push('afterAll2'),
        async () => { throw new Error('afterAll hook error'); },
      );

      try {
        await parent.asyncRun();
      } catch {
        // noop
      }
      expect(calls).toMatchSnapshot();

      expect(console.error).toHaveBeenCalledTimes(3);
      expect(console.error).toHaveBeenCalledWith('Error in beforeEach hook: Error: beforeEach hook error');
      expect(console.error).toHaveBeenCalledWith('Error in afterAll hook: Error: afterAll hook error');

      console.error = error;
    });
  });
});
