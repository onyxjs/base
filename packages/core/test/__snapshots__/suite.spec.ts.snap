// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suite should invoke hooks 1`] = `
Array [
  "beforeAll1",
  "beforeAll2",
  "beforeEach1",
  "beforeEach1",
  "beforeEach2",
  "beforeEach2",
  "afterEach1",
  "afterEach2",
  "afterAll1",
  "afterAll2",
]
`;
